### PR TITLE
Feature/validate repository functions

### PR DESCRIPTION
**Pull request by Thibaud-Babin**
# Add Repository Functions for Request Closure and Kudos System

## Purpose

This PR adds repository-level functions to support request closure and kudos awarding. These functions provide the data layer foundation for the upcoming **Validate Request Screen** feature.

**Related:**
- Issue: #222 
- User Story: #209
- **Part 1 of 3**: Repository Layer → ViewModel Layer → UI Layer

#### `RequestRepository.kt` / `RequestRepositoryFirestore.kt`
-  **New Function**: `closeRequest(requestId: String, selectedHelperIds: List<String>): Boolean`
  - **Purpose**: Closes a request and determines if creator should receive bonus kudos
  - **Returns**: `true` if creator should receive kudos (when helpers selected), `false` otherwise
  - **Validations**:
    - User must be authenticated
    - User must own the request
    - Request status must be OPEN or IN_PROGRESS
    - All selected helpers must be actual request participants
  - **Updates**: Changes request status to COMPLETED

#### `UserProfileRepository.kt` / `UserProfileRepositoryFirestore.kt`
-  **New Function**: `awardKudos(userId: String, amount: Int)`
  - **Purpose**: Awards kudos to a single user
  - **Validations**:
    - Amount must be positive
    - Amount cannot exceed `MAX_KUDOS_PER_TRANSACTION`
    - User must exist in database
  - **Updates**: Both public and private profiles atomically using `FieldValue.increment()`

-  **New Function**: `awardKudosBatch(awards: Map<String, Int>)`
  - **Purpose**: Awards kudos to multiple users in a single atomic transaction
  - **Validations**:
    - All amounts must be positive
    - Individual amounts cannot exceed `MAX_KUDOS_PER_TRANSACTION`
    - Total kudos cannot exceed `MAX_KUDOS_PER_TRANSACTION`
    - All users must exist before committing
  - **Atomicity**: Uses Firestore batch operation - all updates succeed or all fail
  - **Updates**: Both public and private profiles for all users

#### `KudosConstants.kt` (New File)
```kotlin
object KudosConstants {
    const val KUDOS_PER_HELPER = 50
    const val KUDOS_FOR_CREATOR_RESOLUTION = 20
    const val MAX_KUDOS_PER_TRANSACTION = 1000
}
```

### Error Handling

**New Exception Classes:**

`RequestClosureException` (sealed class):
- `InvalidStatus(status)` - Request in wrong status (COMPLETED, CANCELLED, ARCHIVED)
- `UserNotHelper(userId)` - Selected user not in request participants
- `UpdateFailed(requestId, cause)` - Firestore operation failed

`KudosException` (sealed class):
- `InvalidAmount(amount)` - Amount invalid (zero, negative, or exceeds limit)
- `UserNotFound(userId)` - User profile doesn't exist
- `TransactionFailed(userId, cause)` - Firestore operation failed

### Return Value Design
`closeRequest` returns `Boolean` instead of throwing on "no helpers":
-  Allows request closure without helpers (valid use case)
-  Caller decides kudos award logic based on return value
-  Cleaner separation of concerns

### Run Integration Tests
```bash
# Ensure Firebase emulator is installed
firebase setup:emulators:firestore

# Run tests
./gradlew connectedAndroidTest --tests "*RequestRepositoryFirestoreTest*"
./gradlew connectedAndroidTest --tests "*UserProfileRepositoryFirestoreTest*"
```

### Manual Testing via Emulator
```kotlin
// Close a request
val shouldAwardCreator = repository.closeRequest("request-id", listOf("helper1", "helper2"))

// Award kudos to single user
repository.awardKudos("user-id", 50)

// Batch award kudos
repository.awardKudosBatch(mapOf(
    "helper1" to 50,
    "helper2" to 50,
    "creator" to 20
))
```

##  Next Steps

**Part 2 (Next PR)**: `ValidateRequestViewModel` implementation
- Consumes these repository functions
- Implements state machine for request validation flow
- Handles helper selection and kudos calculation

**Part 3 (Following PR)**: Validate Request Screen UI
- User interface for request closure
- Helper selection checkboxes
- Confirmation dialog with kudos preview

### Key Points
1. **Atomicity**: Batch operations guarantee all-or-nothing semantics
2. **Security**: All operations validate authentication and ownership
3. **Testing**: Comprehensive integration tests with Firebase emulator
4. **Error Handling**: Clear exception types with descriptive messages
5. **Validation**: Multiple layers prevent invalid data

### Review Focus Areas
-  Security: Authentication and authorization checks
-  Data integrity: Atomic operations and rollback behavior
-  Error handling: Exception types and messages
-  Test coverage: Edge cases and failure scenarios
-  Performance: Batch operations vs. individual updates

##  Documentation

- KDoc comments added to all new functions
- Exception classes document failure scenarios
- Test files demonstrate usage patterns
- Constants clearly define business rules

---

## AI Use Disclosure
AI was used to help with test implementation and documentation.

### CODE REVIEW

File: app/src/main/java/com/android/sample/model/profile/UserProfileRepositoryFirestore.kt [hunk 1]
```diff
@@ -155,4 +158,84 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
...
+      throw e
+    } catch (e: Exception) {
+      throw KudosException.TransactionFailed(userId, e)
+    }
+  }
+
+  override suspend fun awardKudosBatch(awards: Map<String, Int>) {
+    // Validate all amounts first
+    awards.forEach { (userId, amount) ->
+      if (amount <= 0) {
```
Comment (Funoma) [inline app/src/main/java/com/android/sample/model/profile/UserProfileRepositoryFirestore.kt]: Maybe put this 0 as a constant

File: app/src/androidTest/java/com/android/sample/utils/BaseEmulatorTest.kt [hunk 2]
```diff
@@ -61,4 +62,20 @@ abstract class BaseEmulatorTest {
     FirebaseEmulator.clearAuthEmulator()
     FirebaseEmulator.signOut()
   }
+
+  // Add this to your BaseEmulatorTest class
+
+  /** Creates a new user with the given email and password, then signs them in. */
+  protected suspend fun createAndSignInUser(email: String, password: String) {
```
Comment (Funoma) [inline app/src/androidTest/java/com/android/sample/utils/BaseEmulatorTest.kt]: Nice function, it will be useful !

File: app/src/androidTest/java/com/android/sample/model/request/RequestRepositoryFirestoreTest.kt
Comment (Funoma) [inline app/src/androidTest/java/com/android/sample/model/request/RequestRepositoryFirestoreTest.kt]: Nice test !

**Comment by Thibaud-Babin** [comment]
> It's a really good PR, and I accept it now, but I just want you to answer my question about awardKudos ( I probably don't understand something). Thank you !

Thank your for your review big Guy, I'll answer it for you right now ! I also removed the maigc values you noticed

File: app/src/androidTest/java/com/android/sample/utils/BaseEmulatorTest.kt (hunk 2 above)
Comment (Thibaud-Babin) [reply app/src/androidTest/java/com/android/sample/utils/BaseEmulatorTest.kt]: thank you 🫢
